[9.x] Posibility to calculate backoff value from total of exceptions, or from total of attempts

MartinPham opened the pull request:
<!--
Please only send a pull request to branches which are currently supported: https://laravel.com/docs/releases#support-policy 

If you are unsure which branch your pull request should be sent to, please read: https://laravel.com/docs/contributions#which-branch

Pull requests without a descriptive title, thorough description, or tests will be closed.

In addition, please describe the benefit to end users; the reasons it does not break any existing features; how it makes building web applications easier, etc.
-->

#39976 

This PR ~~makes the Worker to prefer to calculate the backoff value from total of exceptions (when the backoff is configured as an array), and fallback to the total of attempts.~~ allows us to configure how to calculate backoff value, based on total of attempts (default), or on total of exceptions.

## Situation
- A Job is configured with backoff as an array `[1, 10]`

### Current behavior
- 1st attempt, the job calls `release()`.
- (After `0` seconds) 2nd attempt, the job throws an exception.
- (After `10` seconds) 3rd attempt, the job throws an exception.
- (After `10` seconds) 4th attempt, the job throws an exception.
...
Notice that on the 2nd attempt, the backoff was `10` seconds instead of `1`

### Proposed solution
1) Introducing `$job->backoffMode`: 
- `Illuminate\Queue\Jobs\Job::BACKOFF_FROM_ATTEMPTS` : (Default) Calculate backoff based on total of attempts
- `Illuminate\Queue\Jobs\Job::BACKOFF_FROM_EXCEPTIONS `: Calculate backoff based on total of exceptions

**Example**
```php
<?php

namespace App\Jobs;

use Exception;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldBeUnique;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Illuminate\Queue\Jobs\Job;

class ExampleJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    public $tries = 10;
    public $maxExceptions = 5;
    
    // Define as class variable
    // public $backoffMode = Job::BACKOFF_FROM_EXCEPTIONS;

    public function __construct()
    {
    }

    public function backoff()
    {
        return [1, 5];
    }

    // Define as class method
    public function backoffMode()
    {
        return Job::BACKOFF_FROM_EXCEPTIONS;
    }

    public function handle()
    {
        if ($this->attempts() === 1) {
            $this->release();
        } else {
            throw new Exception();
        }
    }
}

```

2) New behavior (BACKOFF_FROM_EXCEPTIONS)
- 1st attempt, the job calls `release()`.
- (After `0` seconds) 2nd attempt, the job throws an exception.
- (After `1` seconds) 3rd attempt, the job throws an exception.
- (After `10` seconds) 4th attempt, the job throws an exception.
...
Notice that on the 2nd attempt, the backoff was `1` seconds, follows the backoff array

Commit:
Calculate the backoff value from total of exceptions

Commit:
formatting

MartinPham commented:
I create the PR on 9.x following the last comment on the same PR on 8.x


If it gets approved & merged, I will make a PR to update the docs

taylorotwell commented on src/Illuminate/Queue/Worker.php:
```diff
@@ -538,19 +573,23 @@ protected function markJobAsFailedIfWillExceedMaxAttempts($connectionName, $job,
-        }
+        $jobExceptionsKey = $this->getJobTotalExceptionsKey($job);
 
-        if ($maxExceptions <= $this->cache->increment('job-exceptions:'.$uuid)) {
```
A race condition has been introduced here. Previously, the `increment` operation here was atomic for some cache systems (Redis, Database, etc.)... but now it has been changed to read, perform some PHP code, and then increment in certain conditions, meaning there could be a race condition between the read and the increment that could lead to unexpected flow.

MartinPham replied:
@taylorotwell Thanks for your review! I've changed the way it increases and compares, to avoid the race condition. Could you please take a look at this?

Commit:
Avoid race condition when increasing total exceptions

Commit:
formatting

Commit:
formatting

humaneguy commented:
@MartinPham Could you provide some unit tests?

MartinPham commented:
@humaneguy sure! I'm gonna do now. Thanks

Commit:
introduce $backoffMode, default to total of attempts - same behavior like before

Commit:
formatting

Commit:
default backoff mode for test

Commit:
add test cases
- calculate backoff with total of attempts
- calculate backoff with total of exceptions

Commit:
no message

Commit:
formatting

Commit:
correct interface's implementation

Commit:
proper cache

Commit:
formatting

Commit:
formatting

MartinPham commented:
@humaneguy 
- I've improved the behavior, so now it's compatible with previous configuration. We can config to change to behavior when we need.
- I've also added test cases against 2 configurations.
Thanks!

humaneguy commented:
@MartinPham Great

humaneguy commented on src/Illuminate/Queue/Worker.php:
```diff
@@ -528,6 +529,55 @@ protected function markJobAsFailedIfWillExceedMaxAttempts($connectionName, $job,
+     */
+    private function getJobTotalExceptions($job)
+    {
+        if (! $this->cache || is_null($job->uuid())) {
```
This check is already done in the ```setupJobTotalExceptionsIfMissing``` method. No need repeating it here.

humaneguy commented on tests/Queue/QueueWorkerTest.php:
```diff
@@ -391,6 +430,13 @@ private function workerOptions(array $overrides = [])
+    {
+        $repository = new Repository(new ArrayStore);
+
+        return $repository;
```
No need to return a variable. It could just be a one-liner.

humaneguy commented on src/Illuminate/Queue/Worker.php:
```diff
@@ -597,7 +647,27 @@ protected function calculateBackoff($job, WorkerOptions $options)
-        return (int) ($backoff[$job->attempts() - 1] ?? last($backoff));
+        $backoffMode = $job->backoffMode();
+
+        if ($backoffMode === Job::BACKOFF_FROM_ATTEMPTS) {
```
Since this is the default case, I'd think you should save it for later.
First check if the backoff mode is calculated by exceptions and return its backoff calculation. Then return this default calculation mode later on.
```suggestion
        if ($backoffMode === Job:: BACKOFF_FROM_EXCEPTIONS) {
            // back off from exceptions calculation
         }
        
        // default calculation of backoff
         return (int) ($backoff[$job->attempts() - 1] ?? last($backoff));
```

MartinPham replied:
Thanks, I was thinking about the default calculation mode also. But then I think maybe it’s better to throw an exception  when the backoffMode’s value is unsupported.
What do you think?

humaneguy replied:
I think it's just safer to use the default backoff mode.

MartinPham replied:
ok!

dammy001 commented on src/Illuminate/Queue/Jobs/Job.php:
```diff
@@ -285,6 +285,22 @@ public function backoff()
+     */
+    public function backoffMode()
+    {
+        return $this->payload()['backoffMode'] ?? self::BACKOFF_FROM_ATTEMPTS;
```
The variable is a not static variable so I think using self is okay

humaneguy replied:
It allows for late static binding. I think it's safer that way.

dammy001 commented on src/Illuminate/Queue/Worker.php:
```diff
@@ -528,6 +529,55 @@ protected function markJobAsFailedIfWillExceedMaxAttempts($connectionName, $job,
+     * Init the job's total number of exceptions if it's missing.
+     *
+     * @param  \Illuminate\Contracts\Queue\Job  $job
+     * @return null
```
I think this should be `null|void`

Commit:
resolve reviewer's comments

MartinPham commented:
Thanks a lot @humaneguy @dammy001!! I've pushed the improvements following your reviews!

taylorotwell commented:
I think I'm just going to hold off making any changes here for the time being.

humaneguy commented:
Are you revisiting this sometime soon? @taylorotwell
